addons/account_full_reconcile/__init__.py

```diff
@@ -1,108 +1,325 @@
+            False
+        )
+        # 1. Reconcile equal amounts:
+        for debit_record in debit_lines:
+            for credit_record in credit_lines:
+                if debit_record.amount_residual == \
+                        -credit_record.amount_residual and \
+                        debit_record.amount_residual > 0:
+                    reconcile_records(
+                        cr, debit_record, credit_record, full_reconcile_id
+                    )
+                    break
+        # 2. Reconcile unequal amounts:
+        current_debit = 0
+        current_credit = 0
+        last_debit = len(debit_lines) - 1
+        last_credit = len(credit_lines) - 1
+        while True:
+            debit_record = debit_lines[current_debit]
+            credit_record = credit_lines[current_credit]
+            if (debit_record.amount_residual > 0 and
+                    credit_record.amount_residual < 0):
+                reconcile_records(
+                    cr, debit_record, credit_record, full_reconcile_id
+                )
+            if current_debit == last_debit and current_credit == last_credit:
+                # Nothing more to reconcile
+                break
+            unexpected_data = True
+            if debit_record.amount_residual <= 0 and \
+                    current_debit < last_debit:
+                current_debit += 1
+                unexpected_data = False
+            if credit_record.amount_residual >= 0 and \
+                    current_credit < last_credit:
+                current_credit += 1
+                unexpected_data = False
+            if unexpected_data:
+                # guard against endless loop due to unexpected/wrong data:
+                # could be caused by credit line with amount_residual > 0
+                # or debit_line with amount_residual < 0.
+                if debit_record.amount_residual < 0.0 or \
+                        credit_record.amount_residual > 0.0:
+                    _logger.error(
+                        "Unexpected data prevented reconcile of lines,"
+                        " debit_line=%s, credit_line=%s" %
+                        (debit_record, credit_record)
+                    )
+                if current_debit < last_debit:
+                    current_debit += 1
+                if current_credit < last_credit:
+                    current_credit += 1
+        # Update amount residual in reconciled records:
+        update_account_move_line(cr, debit_lines, full_reconcile_id)
+        update_account_move_line(cr, credit_lines, full_reconcile_id)
+
+    # Define generator to loop over potentially huge amount of records in
+    # cursor
+    def result_iter(cursor, arraysize=1024):
+        'An iterator that uses fetchmany to keep memory usage down'
+        while True:
+            try:
+                results = cursor.fetchmany(arraysize)
+            except:
+                # Assume no more rows to fetch
```

Comment:
Maybe not all implementations are the same. In the original ecample I followed there always was an exception when there were no more rows. As this did not function I added the test for a falsey result. And left the exception handling there just in case. I agree it might be better to at least log the fact that an exeption occured.